### PR TITLE
[JENKINS-35192] [JENKINS-35244] [JENKINS-35281] Unit tests

### PR DESCRIPTION
**Pull request by alexsomai**
Write unit tests for the features implemented so far.

[[JENKINS-35192]](https://issues.jenkins-ci.org/browse/JENKINS-35192) Unit tests for the exwsAllocate step
[[JENKINS-35244]](https://issues.jenkins-ci.org/browse/JENKINS-35244) Unit tests for the exws step
[[JENKINS-35281]](https://issues.jenkins-ci.org/browse/JENKINS-35281) Unit tests for the exwsAllocate step - downstream job

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateExecution.java [hunk 1]
```diff
@@ -62,15 +62,15 @@ protected ExternalWorkspace run() throws Exception {
...
-            if (physicalPathOnDisk == null) {
-                String message = format("Physical path on disk was not provided in the Jenkins global config for the Disk Pool ID '%s'", diskPoolId);
-                throw new AbortException(message);
-            }
             String diskId = disk.getDiskId();
             if (diskId == null) {
                 String message = format("Disk ID was not provided in the Jenkins global config for the Disk Pool ID '%s'", diskPoolId);
                 throw new AbortException(message);
             }
+            if (physicalPathOnDisk == null) {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateExecution.java]: Was it a bugfix?

Comment (alexsomai) [inline src/main/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateExecution.java]: No, I have just changed the order of null checks.
When I was writing the tests, I though it was more appropriate to check firstly if the diskId is null, and the check if the physicalPathOnDisk is null.
Before, I was checking if the physicalPathOnDisk is null, and afterwards if the diskId is null.

**Comment by alexsomai** [comment]
This PR is ready for review.
CC @oleg-nenashev @martinda

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void missingMasterMountPoint() throws Exception {
+        setupDiskPool(new Disk(DISK_ID, "name", "", "path"));
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertLogContains(format("Mounting point from Master to the disk is not defined for Disk ID '%s', from Disk Pool ID '%s'", DISK_ID, DISK_POOL_ID), upstreamRun);
+    }
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I think it is okay to have the `missingMasterMountPoint` test here for now, but when there is more than one allocation strategy, it may be better to place strategy testing in a test file of its own. At this point, no change is needed IMO.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        Disk disk = new Disk(DISK_ID, "name", "mount", "path");
+        setupDiskPool(disk);
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatusSuccess(upstreamRun);
+        j.assertLogContains(format("Selected Disk ID '%s' from the Disk Pool ID '%s'", DISK_ID, DISK_POOL_ID), upstreamRun);
+        j.assertLogContains(format("The path on Disk is: %s/%s/%d", disk.getPhysicalPathOnDisk(), upstreamRun.getParent().getFullDisplayName(), upstreamRun.getNumber()), upstreamRun);
+    }
+
+    /* ##### Tests for the downstream Job ###### */
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: Not important: my personal preference would have been to split the upstream and the downstream tests into two test files, but this is not a priority for now. The tests do run very fast on my computer, so no issue in terms of execution time. No change needed at this time.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void upstreamJobHasNoActionRegistered() throws Exception {
+        String jobName = "test";
+        createWorkflowJobAndRun(jobName, "echo 'hello world'");
+        createDownstreamJobAndRun(jobName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("The Jenkins job '%s' does not have registered any 'External Workspace Allocate' action", jobName), downstreamRun);
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I think better grammar would be "... hasn't register any 'External ...", but I think the error message could be made more clear to the user.

When I write error messages in my code, I try to tell the user how they got in trouble in the first place, and how they can fix it. When I read this code, I ask myself: the lack of "run action" is a symptom of what? We want to make sure the novice user does not go on a tangent looking for ways to add a "run action", when all they have to do is call `exwsAllocate`. Ok, so let's park this idea and look at some corner cases.

I think there is a corner case where the user has a build history like this:
2. upstream calls `exwsAllocate` but fails **before** it registers the action.
1. upstream does not call `exwsAllocate` and is stable (green).

And another corner case, almost identical to the case above:
2. upstream calls `exwsAllocate` but fails **after** it registers the action
1. upstream does not call `exwsAllocate` and is stable (green)

We want the user to have a build that is stable and that calls `exwsAllocate`, so it has the "action" the code is looking for. Try to write an error message that guides the user in this direction.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void upstreamBuildIsNotPipelineJob() throws Exception {
+        String jobName = "name";
+        FreeStyleBuild build = j.createFreeStyleProject(jobName).scheduleBuild2(0).get();
+        createDownstreamJobAndRun(jobName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("Build '%s' is not a Pipeline job. Can't read the run actions", build), downstreamRun);
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: Not important: I can't seem to find the code that causes this error message in the plugin code, so I guess it comes from the pipeline code?

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    @Test
+    public void missingMasterMountPoint() throws Exception {
+        setUpDiskPool(new Disk(DISK_ID_ONE, "name", "", "path"));
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertLogContains(format("Mounting point from Master to the disk is not defined for Disk ID '%s', from Disk Pool ID '%s'", DISK_ID_ONE, DISK_POOL_ID), upstreamRun);
+    }
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I think it is okay to have the `missingMasterMountPoint` test here for now, but when there is more than one allocation strategy, it may be better to place strategy testing in a test file of its own. At this point, no change is needed IMO.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        createUpstreamJobAndRun();
+
+        Disk allocatedDisk = findAllocatedDisk(disk1, disk2);
+
+        j.assertBuildStatusSuccess(upstreamRun);
+        j.assertLogContains(format("Selected Disk ID '%s' from the Disk Pool ID '%s'", allocatedDisk.getDiskId(), DISK_POOL_ID), upstreamRun);
+        j.assertLogContains(format("The path on Disk is: %s/%s/%d", allocatedDisk.getPhysicalPathOnDisk(), upstreamRun.getParent().getFullDisplayName(), upstreamRun.getNumber()), upstreamRun);
+    }
+
+    /* ##### Tests for the downstream Job ###### */
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: Not important: my personal preference would have been to split the upstream and the downstream tests into two test files, but this is not a priority for now. The tests do run very fast on my computer, so no issue in terms of execution time. No change needed at this time.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    @Test
+    public void upstreamJobHasNoActionRegistered() throws Exception {
+        String jobName = randomAlphanumeric(10);
+        createWorkflowJobAndRun(jobName, "echo 'hello world'");
+        createDownstreamJobAndRun(jobName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("The Jenkins job '%s' does not have registered any 'External Workspace Allocate' action. Did you run exwsAllocate step in the upstream job?", jobName), downstreamRun);
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I think better grammar would be "... hasn't register any 'External ...", but I think the error message could be made more clear to the user.

When I write error messages in my code, I try to tell the user how they got in trouble in the first place, and how they can fix it. When I read this code, I ask myself: the lack of "run action" is a symptom of what? We want to make sure the novice user does not go on a tangent looking for ways to add a "run action", when all they have to do is call `exwsAllocate`. Ok, so let's park this idea and look at some corner cases.

I think there is a corner case where the user has a build history like this:
2. upstream calls `exwsAllocate` but fails **before** it registers the action.
1. upstream does not call `exwsAllocate` and is stable (green).

And another corner case, almost identical to the case above:
2. upstream calls `exwsAllocate` but fails **after** it registers the action
1. upstream does not call `exwsAllocate` and is stable (green)

We want the user to have a build that is stable and that calls `exwsAllocate`, so it has the "action" the code is looking for. Try to write an error message that guides the user in this direction.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        j.assertBuildStatusSuccess(upstreamRun);
+        j.assertBuildStatusSuccess(downstreamRun);
+        j.assertLogContains("WARNING: Both 'upstream' and 'diskPoolId' parameters were provided. " +
+                "The 'diskPoolId' parameter will be ignored. The step will allocate the workspace used by the upstream job.", downstreamRun);
+        j.assertLogContains(format("Selected Disk ID '%s' from the Disk Pool ID '%s'", DISK_ID_ONE, DISK_POOL_ID), downstreamRun);
+        j.assertLogContains(format("The path on Disk is: %s/%s/%d", disk.getPhysicalPathOnDisk(), upstreamName, upstreamRun.getNumber()), downstreamRun);
+    }
+
+    @Test
+    public void upstreamJobRegisteredMultipleActions() throws Exception {
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: It is very good that you thought about this case.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsStepTest.java [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+
+        j.assertBuildStatusSuccess(run);
+        j.assertLogContains("Searching for disk definitions in the External Workspace Templates from Jenkins global config", run);
+        j.assertLogNotContains("Searching for disk definitions in the Node config", run);
+        j.assertLogContains(format("Running in %s/%s/%s/%d",
+                allocatedDisk.getMasterMountPoint(), allocatedDisk.getPhysicalPathOnDisk(), run.getParent().getName(), run.getNumber()), run);
+        // The text written to file should be printed twice on the console output (when writing and when reading the file)
+        assertThat(countMatches(JenkinsRule.getLog(run), TEXT), is(2));
+    }
+
```
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsStepTest.java]: I can think of one more test: a test with two different workflow jobs using the same workspace. Not so much to prove that it works, but to prevent future modifications to the plugin from breaking it without noticing.

**Comment by martinda** [comment]
From my point of view, just two items need to be addressed, and this pull request is ready.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 2 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I was thinking to add test classes for each allocation strategy when we'll have multiple implementations.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 3 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: This is something I'll take into consideration when multiple features will be implemented for the `exwsAllocate` step.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 5 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: The current branch version doesn't have this test anymore.
This was a test case for an older implementation: when `exwsAllocate` step was registering the run actions on the `flowNode` object. Now it registers the run action on the `run` object.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 4 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: Please let me know if you find future grammar mistakes, since my english has some flaws :)

Regarding the corner cases:
> .2. upstream calls exwsAllocate but fails **before** it registers the action.

and 

> .2. upstream calls exwsAllocate but fails **after** it registers the action

If the build fails, no matter at which point, that build number will not be taken into consideration anyway.

But I can improve this error message for the case when the upstream build is green, but the user didn't call `exwsAllocate` step.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 6 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I was thinking to add test classes for each allocation strategy when we'll have multiple implementations.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 8 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: Please let me know if you find future grammar mistakes, since my english has some flaws :)

Regarding the corner cases:

> .2. upstream calls exwsAllocate but fails before it registers the action.

and

> .2. upstream calls exwsAllocate but fails after it registers the action
If the build fails, no matter at which point, that build number will not be taken into consideration anyway.

But I can improve this error message for the case when the upstream build is green, but the user didn't call exwsAllocate step.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 7 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: This is something I'll take into consideration when multiple features will be implemented for the exwsAllocate step.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsStepTest.java (hunk 10 above)
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsStepTest.java]: Yes, good point. I'll add that.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java [hunk 11]
```diff
@@ -159,11 +159,13 @@ public void upstreamJobHasNoStableBuild() throws Exception {
...
         String jobName = randomAlphanumeric(10);
-        createWorkflowJobAndRun(jobName, "echo 'hello world'");
+        upstreamRun = createWorkflowJobAndRun(jobName, "echo 'hello world'");
         createDownstreamJobAndRun(jobName);
 
         j.assertBuildStatus(FAILURE, downstreamRun);
-        j.assertLogContains(format("The Jenkins job '%s' does not have registered any 'External Workspace Allocate' action. Did you run exwsAllocate step in the upstream job?", jobName), downstreamRun);
+        j.assertLogContains(format("This downstream job has to allocate the same external workspace used by the upstream job. " +
+                "To do so, the exwsAllocate step must be called in the upstream job. " +
+                "Please call the exwsAllocate step in the upstream job, because the build '%s' doesn't have such calls.", upstreamRun), downstreamRun);
```
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I have improved this error message. The actual change is in #8, this is only the test associated with it.

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsStepTest.java [hunk 12]
```diff
@@ -215,6 +215,28 @@ public void sharedWorkspaceBetweenTwoDifferentNodesWithTemplate() throws Excepti
         assertThat(countMatches(JenkinsRule.getLog(run), TEXT), is(2));
     }
 
+    @Test
+    public void twoDifferentJobsUsingTheSameWorkspace() throws Exception {
```
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsStepTest.java]: I have added one more test case that @martinda suggested.

**Comment by oleg-nenashev** [comment]
Looks good to me. I would merge it in order to get some testing in the new PRs.
Maybe there is some minor changes and extra tests required, but should work fine for the current phase

**Comment by oleg-nenashev** [comment]
BTW a good practice is to summarize changes and to add issue hyperlinks to the PR summary

File: src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java (hunk 11 above)
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java]: I commented on the actual change in PR #8 . I find it difficult to come up with a good error message for this error condition, but please see my comment, I think I have nailed it.

**Comment by alexsomai** [comment]
This PR had to be created against `master` branch. Closing in favor of #14.
